docs/src/pages/docs/guides/ssr.md

```diff
@@ -5,59 +5,175 @@ title: SSR & Next.js
+// pages/posts.jsx
+import { makeQueryCache } from 'react-query'
+import { dehydrate } from 'react-query/hydration'
+
+export async function getStaticProps() {
+  const queryCache = makeQueryCache()
+
+  await queryCache.prefetchQuery('posts', getPosts)
+
+  return {
+    props: {
+      initialQueries: dehydrate(queryCache)
+    }
+  }
+}
 
-- If you import and use the global `queryCache` directly, queries are not cached during SSR to avoid leaking sensitive information between requests.
-- If you create a `queryCache` manually with `makeQueryCache`, queries will be cached during SSR. Make sure you create a separate cache per request to avoid leaking data.
-- Queries rendered on the server will by default use the `initialData` of an unfetched query. This means that by default, `data` will be set to `undefined`. To get around this in SSR, you can either pre-seed a query's cache data using the `config.initialData` option:
+function Posts() {
+  // This useQuery could just as well happen in some deeper child to
+  // the "Posts"-page, data will be available immediately either way
+  const { data } = useQuery('posts', getPosts)
 
-```js
-const queryInfo = useQuery('todos', fetchTodoList, {
-  initialData: [{ id: 0, name: 'Implement SSR!' }],
-})
+  // This query was not prefetched on the server and will not start
+  // fetching until on the client, both patterns are fine to mix
+  const { data: otherData } = useQuery('posts-2', getPosts)
 
-// data === [{ id: 0, name: 'Implement SSR!'}]
+  // ...
+}
 ```
 
-Or, alternatively you can just destructure from `undefined` in your query results:
+As demonstrated, it's fine to prefetch some queries and let some fetch on the client. This means you can control what content server renders or not by adding or removing `prefetchQuery` for a specific query.
+
+### Integrating with custom SSR solutions or other frameworks
+
+Since there are many different possible setups for SSR, it's hard to give a detailed guide for each (contributions are welcome!). Here is a thorough high level overview:
+
+**Server side**
+
+> Note: The global `queryCache` you can import directly from 'react-query' does not cache queries on the server to avoid leaking sensitive information between requests.
+
+- Prefetch data
+  - Create a `prefetchQueryCache` specifically for prefetching by calling `const prefetchQueryCache = makeQueryCache()`
+  - Call `prefetchQueryCache.prefetchQuery(...)` to prefetch queries
+  - Dehydrate by using `const initialQueries = dehydrate(prefetchQueryCache)`
+- Render
+  - Wrap the app in `<ReactQueryCacheProvider>` from `'react-query/hydration'` and pass in `initialQueries`
+    - This makes sure a separate `queryCache` is created specifically for rendering
+    - **Do not** pass in the `prefetchQueryCache` from the last step, the server and client both needs to render from the dehydrated data to avoid React hydration mismatches. This is because queries with errors are excluded from dehydration by default.
```

Comment:
This still feels a bit shaky. We could require users to filter themselves with:
```
prefetchCache.removeQueries(query => query.state.isError)
```
Or do it automatically in the `dehydrate` function?

Reply:
Even with that approach, the one documented now will still work too, so this comes down to what we should recommend as an approach in the docs. I think requiring `.removeQueries` in some cases but not others is equally shaky and confusing and would rather just recommend a single approach for Next and custom SSR, even with the above caveat.

I'm very open to adding `.removeQueries` later though, as well as changing the docs if we see this becoming too confusing for users.

As for making `dehydrate` remove it automatically, I really don't want to make `dehydrate` mutative and destructive for several reasons, main one being it can also be used with localstorage where you wouldn't want that.

Reply:
Yes good point, the `dehydrate` function should indeed not be mutative. But people may also expect the `dehydrate` function to dehydrate everything by default. The `removeQueries` method already exists and using it would make it very clear to users on what is happening:
1. Prefetch
2. Filter out anything you don't want
3. Dehydrate
4. Render

There would be no implicit behaviour or juggling needed.

Reply:
Oh, look at that! I had totally missed that the function already existed, nice! :)

There is definitely the case where you don't want to remove ahead of time, but only on the dehydrate step (for localstorage), but there is `shouldDehydrate` for that.

I am starting to warm to this a little bit, I still think there is a big case to be made for `dehydrate` filtering by default since that's the major usecase and we want to support that as easily as possible, but there really a big point to be made for no implicit behaviour as well.

For custom SSR, you might be able to set it up in a single place. The sad part though is that in Next, you would have to call `removeQueries` or provide a `shouldDehydrate` in every single page where you fetch data, and if you forget, you can quite possibly get errors in production (because no queries failed in dev). I think that's a major drawback.

Reply:
Since `removeQueries` already exists and docs can be changed, what we need to decide is ultimately if `dehydrate` should only hydrate successful queries by default.

Reply:
The more I think about this the more I think we should filter unsuccessful queries by default. It's easier to get started with and is the safe default. Often when you develop you only have successful queries, so it's easy to end up with pretty bad errors in prod if you forget to add removeQueries somewhere (as in, the server might crash).

I'll go ahead and resolve this.